File: src/nougen_shards/structured.py
"""
Structured Output Helpers for NouGenShards.
Handles local JSON validation and repair.
"""
import json
import re
from typing import Any, Tuple, List, Optional, Dict

def parse_json_content(content: str) -> Dict[str, Any]:
    """
    Attempts to parse JSON from a string, handling common LLM formatting issues.
    """
    content = content.strip()
    
    # 1. Direct Parse
    try:
        parsed = json.loads(content)
        if isinstance(parsed, dict):
            return parsed
    except json.JSONDecodeError:
        pass
        
    # 2. Extract from Markdown Code Block
    match = re.search(r"```(?:json)?\s*(.*?)\s*```", content, re.DOTALL)
    if match:
        try:
            parsed = json.loads(match.group(1))
            if isinstance(parsed, dict):
                return parsed
        except json.JSONDecodeError:
            pass
            
    # 3. Extract from first { to last }
    start = content.find('{')
    end = content.rfind('}')
    if start != -1 and end != -1:
        try:
            parsed = json.loads(content[start:end+1])
            if isinstance(parsed, dict):
                return parsed
        except json.JSONDecodeError:
            pass
            
    raise ValueError("Failed to parse JSON from content.")

def validate_against_schema(data: Dict[str, Any], schema: Dict[str, Any]) -> Tuple[bool, List[str]]:
    """
    Minimal schema validation fallback.
    Checks required fields and basic types.
    """
    errors: List[str] = []
    
    # Check Required Fields
    required = schema.get("required", [])
    for field in required:
        if field not in data:
            errors.append(f"Missing required field: {field}")
            
    # Check Types (Primitive)
    properties = schema.get("properties", {})
    for field, val in data.items():
        if field in properties:
            expected_type = properties[field].get("type")
            if expected_type == "string" and not isinstance(val, str):
                errors.append(f"Field '{field}' expected string, got {type(val).__name__}")
            elif expected_type == "number" and not isinstance(val, (int, float)):
                errors.append(f"Field '{field}' expected number, got {type(val).__name__}")
            elif expected_type == "integer" and not isinstance(val, int):
                errors.append(f"Field '{field}' expected integer, got {type(val).__name__}")
            elif expected_type == "boolean" and not isinstance(val, bool):
                errors.append(f"Field '{field}' expected boolean, got {type(val).__name__}")
            elif expected_type == "array" and not isinstance(val, list):
                errors.append(f"Field '{field}' expected array, got {type(val).__name__}")
            elif expected_type == "object" and not isinstance(val, dict):
                errors.append(f"Field '{field}' expected object, got {type(val).__name__}")
                
    # Check additionalProperties: false
    if schema.get("additionalProperties") is False:
        for field in data:
            if field not in properties:
                errors.append(f"Unexpected additional property: {field}")
                
    return (len(errors) == 0, errors)

File: ts/src/nougen_shards/structured.ts
/**
 * Structured Output Helpers for NouGenShards. (TS mimic of structured.py)
 * Handles local JSON validation and repair.
 */

function _is_plain_object(val: any): val is Record<string, any> {
  return typeof val === "object" && val !== null && !Array.isArray(val);
}

/**
 * Attempts to parse JSON from a string, handling common LLM formatting issues.
 */
export function parse_json_content(content: string): Record<string, any> {
  content = content.trim();

  // 1. Direct Parse
  try {
    const parsed = JSON.parse(content);
    if (_is_plain_object(parsed)) {
      return parsed;
    }
  } catch {
    /* fallthrough */
  }

  // 2. Extract from Markdown Code Block
  const match = content.match(/```(?:json)?\s*([\s\S]*?)\s*```/);
  if (match) {
    try {
      const parsed = JSON.parse(match[1]);
      if (_is_plain_object(parsed)) {
        return parsed;
      }
    } catch {
      /* fallthrough */
    }
  }

  // 3. Extract from first { to last }
  const start = content.indexOf("{");
  const end = content.lastIndexOf("}");
  if (start !== -1 && end !== -1) {
    try {
      const parsed = JSON.parse(content.slice(start, end + 1));
      if (_is_plain_object(parsed)) {
        return parsed;
      }
    } catch {
      /* fallthrough */
    }
  }

  throw new Error("Failed to parse JSON from content.");
}

function _type_name(val: any): string {
  if (val === null) return "NoneType";
  if (Array.isArray(val)) return "list";
  switch (typeof val) {
    case "string":
      return "str";
    case "number":
      return Number.isInteger(val) ? "int" : "float";
    case "boolean":
      return "bool";
    case "object":
      return "dict";
    default:
      return typeof val;
  }
}

/**
 * Minimal schema validation fallback.
 * Checks required fields and basic types.
 */
export function validate_against_schema(
  data: Record<string, any>,
  schema: Record<string, any>,
): [boolean, string[]] {
  const errors: string[] = [];

  // Check Required Fields
  const required: string[] = schema.required ?? [];
  for (const field of required) {
    if (!(field in data)) {
      errors.push(`Missing required field: ${field}`);
    }
  }

  // Check Types (Primitive)
  const properties: Record<string, any> = schema.properties ?? {};
  for (const [field, val] of Object.entries(data)) {
    if (field in properties) {
      const expected_type = properties[field].type;
      if (expected_type === "string" && typeof val !== "string") {
        errors.push(`Field '${field}' expected string, got ${_type_name(val)}`);
      } else if (expected_type === "number" && typeof val !== "number") {
        errors.push(`Field '${field}' expected number, got ${_type_name(val)}`);
      } else if (expected_type === "integer" && !Number.isInteger(val)) {
        errors.push(`Field '${field}' expected integer, got ${_type_name(val)}`);
      } else if (expected_type === "boolean" && typeof val !== "boolean") {
        errors.push(`Field '${field}' expected boolean, got ${_type_name(val)}`);
      } else if (expected_type === "array" && !Array.isArray(val)) {
        errors.push(`Field '${field}' expected array, got ${_type_name(val)}`);
      } else if (expected_type === "object" && !_is_plain_object(val)) {
        errors.push(`Field '${field}' expected object, got ${_type_name(val)}`);
      }
    }
  }

  // Check additionalProperties: false
  if (schema.additionalProperties === false) {
    for (const field of Object.keys(data)) {
      if (!(field in properties)) {
        errors.push(`Unexpected additional property: ${field}`);
      }
    }
  }

  return [errors.length === 0, errors];
}
